### ISSUE TITLE
Will the code be open-sourced?

### ISSUE THREAD

**Issue by 980202006** [zhangyongmao/VISinger2#1]
Great work! This singing synthesis effect is the best I have heard so far, and the paper is also very scientific. Will you release the code?

**Comment by 980202006** [comment]
I want to use visinger2 for multiple singers, do you have any suggestions?

**Comment by 980202006** [comment]
![image](https://user-images.githubusercontent.com/24452502/203496322-98afb88a-f76e-4473-8f25-2216aa3ec4c3.png)
How does the mel of visinger2 and the hidden layer vector output by the right LR input to the fft block? Whether to use cross attention, or simple addition, or concat?

**Comment by zhangyongmao** [comment]
> Great work! This singing synthesis effect is the best I have heard so far, and the paper is also very scientific. Will you release the code?

Yes, we will release the code later.

**Comment by zhangyongmao** [comment]
> 我想对多个歌手使用visinger2，你有什么建议吗？

I think you can add spk embedding to the input of the frame-level FFT and the duration predictor in the prior encoder.

**Comment by zhangyongmao** [comment]
> ![image](https://user-images.githubusercontent.com/24452502/203496322-98afb88a-f76e-4473-8f25-2216aa3ec4c3.png) How does the mel of visinger2 and the hidden layer vector output by the right LR input to the fft block? Whether to use cross attention, or simple addition, or concat?

I concat the mel and the hidden layer vector directly.

**Comment by 980202006** [comment]
@zhangyongmao What is the kernel size of the cnn of this model?

**Comment by 980202006** [comment]
Are there training-related parameters, such as the maximum audio length, batch_size, the size of the video memory used, and  the num of gpus for training?

**Comment by 980202006** [comment]
Does the downsample block have relevant parameter details, such as directly downsampling to the required multiple, or the downsampling method used?

**Comment by zhangyongmao** [comment]
> ![image](https://user-images.githubusercontent.com/24452502/203507206-44bd5c03-63f6-44cb-8240-0a2b64bcafc5.png) Thank you for your reply! I noticed that the original ddsp has two harmonic implementations, one is streaming_harmonic_synthesis, and the other is harmonic_synthesis. Which implementation does visinger2 use? The text-to-phase problem mentioned by visinger2 is of great research value. In addition, is the inital phase in the figure learnable or fixed in visinger?

We use the harmonic_synthesis in DDSP. The inital phase is fixed.

**Comment by zhangyongmao** [comment]
> @zhangyongmao What is the kernel size of the cnn of this model?

The kernel size of the cnn in the FFT Blocks is 3.

**Comment by zhangyongmao** [comment]
> Does the downsample block have relevant parameter details, such as directly downsampling to the required multiple, or the downsampling method used?

I use Conv1d for downsampling, and the sequence of downsampling multiples is in the reverse order of the sequence of upsampling multiples. For example, if upsample_rates=[8,8,2,2], then the downsample_rates is [2,2,8,8].

**Comment by 980202006** [comment]
Thank you for your reply!I will try it.

**Comment by 980202006** [comment]
Hi, I noticed that harmonic_synthesis has three inputs: energy, f0 and harmonic. But there is no energy in the paper. Did you use the energy originally extracted by ddsp?

**Comment by zhangyongmao** [comment]
> Hi, I noticed that harmonic_synthesis has three inputs: energy, f0 and harmonic. But there is no energy in the paper. Did you use the energy originally extracted by ddsp?

I did not use energy as the input to ddsp because the potential representation z contains the information about energy. I'm also trying to build a multi-speaker model, and I find that flow may be necessary for a multi-speaker model.

**Comment by 980202006** [comment]
Thank you!I will try it.Your reply is very helpful for my reproduction.

**Comment by 980202006** [comment]
@zhangyongmao Can you share your losses? I am training visinger, but I am worried about implementation issues (in order to align the performance of tensorflow, I implemented ddsp here), and I want to know the approximate convergence value of kl loss and so on.

**Comment by zhangyongmao** [comment]
> @zhangyongmao Can you share your losses? I am training visinger2.I am worried about the code bug and I want to know the approximate convergence value of kl loss and so on.

For kl loss, when I don't use flow, I use torch.distributions.kl_divergence, when the weight of kl loss is 1, the kl loss is about 0.2. When using flow, I use the klloss in vits, which is about 1.0.

**Comment by 980202006** [comment]
I still have another question to ask: why use two gradient truncation operations, is it because of the memory problem? I think removing these two gradient truncations helps the discriminator to optimize the acoustic model. @zhangyongmao

**Comment by innnky** [comment]
请问代码大概什么时候会公开呢？

**Comment by zhangyongmao** [comment]
> @zhangyongmao Thanks for your info!! 1-last question for me, how can I prepare my custom English singing dataset I have about 9 hours cut into 5-6 seconds segments , what is the format they should be in ?? like text| "then the music score"? and how can I create this music score, is it manual work or is there a tool to automate it , like if you can tell me the way to prepare the training.txt in English and its exact format , that would be very helpful from you. 2-"Refinegan results seem not good" maybe because Refinegan had 59 hours of different speakers not only 5.2 hours, I would say SiFi-GAN (or avocodo from ncsoft) is better than both hifi-gan and even might compete with Refine , you might try it if you want. Thanks in advance

Sorry I didn't reply in time. I used the annotation format of opencpop, and the format of trainset is: phones, pitchID, note-dur, gt-dur, slur. These annotations should be obtained by automatic labeling and manual correction.

**Comment by zhangyongmao** [comment]
> I still have another question to ask: why use two gradient truncation operations, is it because of the memory problem? I think removing these two gradient truncations helps the discriminator to optimize the acoustic model. @zhangyongmao

Sorry I didn't reply in time. These variables for gradient truncation have specific meaning, not potential variables, so I only calculated a loss and did not use a discriminator to optimize it.

**Comment by zhangyongmao** [comment]
> 请问代码大概什么时候会公开呢？

这个我还在整理代码和预训练模型，可能下周会整理好

**Comment by 980202006** [comment]
Thank you for your reply. Your reply is very helpful for my research and study.

**Comment by innnky** [comment]
> > 请问代码大概什么时候会公开呢？
> 
> 这个我还在整理代码和预训练模型，可能下周会整理好

期待！

**Comment by swagger-coder** [comment]
期待开源ing！

**Comment by yangxipeng** [comment]
给大佬点赞👍

**Comment by splinter21** [comment]
期待开源ing！
